### DISCUSSION TITLE
Using 'google_maps_flutter' tiles with 'flutter_map'

### DISCUSSION THREAD

**Discussion by Ahmed-gubara** [fleaflet/flutter_map#1158] [Show and tell]
Hi
Just had some time to play with google_maps_flutter package.
I would like to share a code that worked for me to use GoogleMap Widget as a tile layer on flutter_map

### usage:
````dart
Widget build(BuildContext context) {
    return Scaffold(
        body: FlutterMap(
      options: MapOptions(
        center: LatLng(51.5, -0.09),
        maxZoom: 21,
        interactiveFlags:
            InteractiveFlag.pinchZoom | InteractiveFlag.doubleTapZoom | InteractiveFlag.drag | InteractiveFlag.flingAnimation | InteractiveFlag.pinchMove,
        zoom: 13.0,
      ),
      children: [
        GoogleMapTileLayerWidget(options: GoogleMapTileLayerOptions(useAndroidViewSurface: true)),
        MarkerLayerWidget(
            options: MarkerLayerOptions(
          markers: [
            Marker(width: 80.0, height: 80.0, point: LatLng(51.5, -0.09), builder: (ctx) => const FlutterLogo()),
          ],
        )),
      ],
    ));
  }
````
There is a slight delay while panning, but it can be reduced greatly with **``useAndroidViewSurface: true``**

**Comment by JaffaKetchup** [comment]
Looks interesting, could you make this a plugin?

**Comment by Ahmed-gubara** [comment]
@JaffaKetchup 
Once i get it to work properly i will make it a plugin.  currently it's still far from optimal.

I tried using it in my app.
There is delay and lag while panning, not visible on an android 11 device, but clearly visible on an android 6 device. 

Both are octa-core CPU with ARM Cortex-A53 cores.

##### _Vehicles are represented using MarkerLayerWidget from flutter_map_
#### Redmi 4x (Snapdragon 435) android 11 Custom ROM

https://user-images.githubusercontent.com/29988012/151365885-21d7fbde-1489-45d8-a845-275912d4d719.mp4



#### Huawei G9 Lite (HiSilicon Kirin 650) android 6 EMUI

https://user-images.githubusercontent.com/29988012/151363910-d137a575-6580-4e7b-91ed-53194822ccad.mp4


And there are other ways for this to be implemented.
1- We can use GoogleMap Widget with `​liteModeEnabled: true` to reduce performance cost

2- We can use googleMapController.takeSnapshot() which will return a Future<Uint8List?> representing the image **bytes** of the map.
That we can use with ``Image.memory(Uint8List bytes)`` as tiles.

I think that is similar to how Leaflet handle google maps tiles, where it uses a google map instance to fetch tiles, and rip it from the DOM tree to use on Leaflet, hopefully this complies with Google's ToS.

**Comment by JaffaKetchup** [comment]
Not sure about compliance with Google ToS, they probably aren't happy about using their tiles without their SDK?

Either option seems good, ultimately it's your decision! If I had to pick, I would probably pick option 2, because it just feels nicer somehow, and less of a workaround.

Thanks for sharing :)

**Comment by ibrierley** [comment]
As I "think" (haven't had chance to dig into it behind the scenes) this is kind of still using the Google Maps SDK, it may actually be fine with their ToS.

**Comment by thomashrabe** [comment]
Does this still work? @Ahmed-gubara

**Comment by Ahmed-gubara** [comment]
@thomashrabe 
Yes, i'm actually using it in an App.


It doesn't perform well on android versions older than android 10 (SDK 29). 
That is something related to flutter using better android platform integration starting with android 10.
[flutter hybrid-composition](https://docs.flutter.dev/platform-integration/android/platform-views#hybrid-composition)

So maybe make a fallback to open street map on older android versions.

```dart
Widget myTileLayer() {
  int androidSdk = getAndroidVersion(); // using device_info_plus
  if (androidSdk >= 29) { // >= Android 10
    return GoogleMapTileLayer(
      mapType: MapType.hybrid,
    );
  }
  else
  {
    return TileLayer(
      urlTemplate: 'https://tile.openstreetmap.org/{z}/{x}/{y}.png',
      userAgentPackageName: 'com.example.app',
    );
  }
}
```

I've updated the gist to the code that i'm currently using.
[google_tile_layer_widget.dart](https://gist.github.com/Ahmed-gubara/ebfb48dbdbff3dc7051ff0998d3e7048#file-google_tile_layer_widget-dart)

The same code will work with [apple_maps_flutter](https://pub.dev/packages/apple_maps_flutter) as the api is almost identical.
```dart
GoogleMap -> AppleMap
GoogleMapController -> AppleMapController
``` 
And it work very well on IOS devices .

**Comment by thomashrabe** [comment]
Thank you for the update. However, I tried it and it seems now that your example is lagging behind. The source file does not contain GoogleMapTileLayerOptions anymore. I think they were in there before?

**Reply by Ahmed-gubara** [reply]
Fixed the example.

GoogleMapTileLayerOptions was removed from the code.

**Reply by thomashrabe** [reply]
Thank you! I can see the map now, but it seems that it does not scroll. iOS 14 simulator. Do I have to do add anything to the FlutterMap to handle gestures and forward these to the GoogleTileLayer?

**Reply by Ahmed-gubara** [reply]
Just updated the gist and tested with IOS simulator, the code should work normally. 
and simplified the example.

It was a simple Container(color: Colors.transparent) at the end of the stack to make gestures captured by flutter_map.

**Reply by mohyM** [reply]
@Ahmed-gubara thanks for ur GoogleMapTileLayer but after i added markers on the map it when zoom the map it not moving correctly with map

https://github.com/fleaflet/flutter_map/assets/12228546/039db732-2f07-4849-9320-0a87549caba8

please see this video

**Reply by Ahmed-gubara** [reply]
@mohyM 
Unfortunately that is the zoom limit supported by google map itself. [CameraUpdateFactory](https://developers.google.com/android/reference/com/google/android/gms/maps/CameraUpdateFactory#parameters_4)

The only way is to cluster markers that are overlapping.
using any plugin supported by FlutterMap, see [marker clustering](https://docs.fleaflet.dev/plugins/list#marker-clustering)


and also:
As i see from the video, you didn't disable map rotation as set by the example.
i have found a way to allow google map to rotate correctly without rotating labels and updated the code.
kindly try the updated code and tell me if it's working correctly for you.

**Reply by mohyM** [reply]
@Ahmed-gubara hi ahmed do u think u can update "GoogleMapTileLayer" class u created before to work fine with flutter_map version > 6.0.0

**Reply by Ahmed-gubara** [reply]
@mohyM 
[google_tile_layer_widget.dart](https://gist.github.com/Ahmed-gubara/2d191bcf3e426521f85f5bf44a071cd7)
This one is for flutter_map v6.
